File: Python/HPCurrency.py

# This class takes in amounts of Harry Potter currency and serves as a base class
class HPCurrency:

    usrGal = 0
    usrSick = 0
    usrKnut = 0

    def __init__(self, usrG, usrS, usrK):
        self.usrGal = usrG
        self.usrSick = usrS
        self.usrKnut = usrK

    def currency(self):
        print("There are two types of conversion into USD")

# This is a child class that converts HP money into USD using a Reddit calculation
class HPRedditConversion(HPCurrency):

    def __init__(self, HPCurrency):
        self.usrGal = HPCurrency.usrGal * 25
        self.usrSick = HPCurrency.usrSick * 1.5
        self.usrKnut = HPCurrency.usrKnut * 0.05

    def currency(self):
        print("According to Reddit:\nUSD In Gals: {}\nUSD in Sicks: {}\nUSD in Knuts: {}".format(self.usrGal,
                                                                           self.usrSick,
                                                                           self.usrKnut))

# This is a child class that converts HP Currency into USD using a Wikia calculation
class HPWikiaConversion(HPCurrency):

    def __init__(self, HPCurrency):
        self.usrGal = HPCurrency.usrGal * 7.35
        self.usrSick = HPCurrency.usrSick * 0.46
        self.usrKnut = HPCurrency.usrKnut * 0.02

    def currency(self):
        print("According to the Wikia:\nUSD In Gals: {}\nUSD in Sicks: {}\nUSD in Knuts: {}".format(self.usrGal,
                                                                           self.usrSick,
                                                                           self.usrKnut))

def conversion():
    inputGal = 0.0
    inputSick = 0.0
    inputKnut = 0.0

    print("How many golden Galleons do you have?")
    inputGal = float(input())

    print("How many silver Sickles do you have?")
    inputSick = float(input())

    print("How many Bronze Knuts do you have?")
    inputKnut = float(input())

    Money = HPCurrency(inputGal, inputSick, inputKnut)

    print("Input Reddit for Reddit conversion, Wikia conversion is the default")

    usrChoice = input().upper()

    if usrChoice == "REDDIT":
        Conversion = HPRedditConversion(Money)
        Conversion.currency()
    else:
        Conversion = HPWikiaConversion(Money)
        Conversion.currency()

File: new_cpp/HPCurrency.cpp
#include <iostream>

using namespace std;

// Base class from which the child classes will inherit their members and features
class HPCurrency {
    public:

        // Gets the amount of Gold Galleons the user has as an integer  
        int getGals(void) {
        
            return galsCount;
        }

       // Sets the amount of Gold Galleons the user has as an integer
        void setGals(int g) {
            
            galsCount = g;
        }

        // Gets the amount of Silver Sickles the user has
        int getSick(void) {

            return sickCount;
        }

        // Sets the amount of Silver Sickles a user has
        void setSick(int s) {
         
            sickCount = s;

        }

        // Gets the amount of Bronze Knuts a user has
        int getKnut(void) {

            return knutCount;
        
        }

        // Sets the amount of Bronze Knuts a user has
        void setKnut(int k) {

            knutCount = k;

        }

        // Provides a way to convert the currency into USD in the derivative/child classes
        float convert() {
            return (getGals() + getSick() + getKnut());
        }

    protected:
        // Instantiates the values of each currency in a limited scope
        int galsCount = 0;
        int sickCount = 0;
        int knutCount = 0;
};

// Derivative/child class of HP Currency that uses the Reddit conversion rate for converting HP currency into USD
class RedditHPCurrencyConversion : public HPCurrency {
     public:
         // Converts the HP Currency using Reddit's conversion rates into USD and returns the value as a decimal/float
         float convert() {
             return (getGals() * 25) + (getSick() * 1.5) + (getKnut() * 0.05);
         }
};

// Derivative/child class of HP Currency that uses a fan Wikia conversion rate for converting HP currency into USD
class WikiaHPCurrencyConversion : public HPCurrency {
    public:
        // Converts the HP Currency using a fan Wikia's conversion rates into USD and returns the value as a decimal/float
        float convert() {
            return (getGals() * 7.35) + (getSick() * 0.46) + (getKnut() * 0.02);
        }
};

int main() {
    // instantiate the variables for user input to 0 for later use
    int usrGals = 0, usrSicks = 0, usrKnuts = 0;

    // Get user input and record their inventory
    cout << "How many Golden Galleons do you have?" << endl;
    cin >> usrGals;
    
    cout << "How many Silver Sickles do you have?" << endl;
    cin >> usrSicks; 

    cout << "How many Bronze Knuts do you have?" << endl;
    cin >> usrKnuts;

    // Default constructor for RedditHPCurrencyConversion, so it does not take any arguments.
    RedditHPCurrencyConversion RedditMoney;

    // The setter methods were inherited from the base/parent class.
    RedditMoney.setGals(usrGals);
    RedditMoney.setSick(usrSicks);
    RedditMoney.setKnut(usrKnuts);

    WikiaHPCurrencyConversion WikiaMoney;

    WikiaMoney.setGals(usrGals);
    WikiaMoney.setSick(usrSicks);
    WikiaMoney.setKnut(usrKnuts);

    // Output the total conversion from the Reddit and Wikia methods
   cout << "Reddit conversion: " << RedditMoney.convert() << "\n" <<
        "Wikia Conversion: " << WikiaMoney.convert()  << endl;

    return 0;
}
